Nxapi ssl

tstoner opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Extended nxos nxapi support for ssl ciphers and TLS protocols with default behavior of weak ciphers and TLSv1.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
nxos_nxapi

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (nxapi_ssl 5d68f9c346) last updated 2018/07/17 11:32:42 (GMT -400)
  config file = None
  configured module search path = [u'/Users/tmstoner/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/tmstoner/ansible/lib/ansible
  executable location = /Users/tmstoner/ansible/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

tstoner committed:
NXAPI ssl ciphers & protocols default values

tstoner committed:
TLSv1, TLSv1.1, TLSv1.2 and weak cipher support

tstoner committed:
NXOS NXAPI weak/strong cipher & TLSv 1.2, 1.1 & 1.0 support

tstoner committed:
Version checking for strong/weak ciphers & TLS 1.2, 1.1 & 1.0 support

tstoner committed:
Cleaned up erroneously committed changes.

tstoner committed:
Specific NXOS platform checking for nxapi ssl ciphers & protocols

mikewiebe commented on lib/ansible/modules/network/nxos/nxos_nxapi.py:
```diff
@@ -94,6 +94,39 @@
+        configured.  To enable the use of strong ciphers, set the value of
+        this argument to True.
+    required: false
+    default: no
```
Here and for the other new `bool` parameters, I would update the default value to be `False`/`True` instead of `no`/`yes` so that it matches what is set in the argument spec in `main()`

tstoner replied:
The convention in the file was to use yes/no in docs and true/false in argument spec.

tstoner committed:
Fixed ansibot reported errors.

tstoner committed:
Resolved ansibot reported error.

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/74679/3/tests

tstoner committed:
Added network_os_version to mocked up N7K unit test device_info

tstoner commented:
@trishnaguha - Hi Trishna, please merge or provide comments if there are concerns.  Br, TomS.

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_nxapi.py:
```diff
@@ -165,6 +203,20 @@ def map_obj_to_commands(want, have, module):
     send_commands = list()
     commands = dict()
 
+    device_info = get_capabilities(module).get('device_info')
```
`get_capabilities` is called in `check_args` as well. Instead of calling the function twice in the module, please call the function in `main()` as pass the return value as argument to the methods.

tstoner committed:
Calling get_capabilities() once in main and passing results into methods.

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_nxapi.py:
```diff
@@ -161,10 +197,24 @@ def check_args(module, warnings):
 
+    device_info = capabilities.get('device_info')
+    if device_info is None:
+        raise TypeError
```
@tstoner I am curious why you are raising `TypeError` if capabilities cannot be fetched. Given that the new options added in this PR are not the required options for `nxos_nxapi` module, I don't think we want to fail the module if `device_info` is None.
Same for `os_version` and `os_platform`. 
The module can simply ignore the task and pass when these parameters are used with a warning message if capabilities isn't available instead of failing/raising exception.

tstoner committed:
Removed raising exceptions when platform capabilities return None
per reviewers request. Skipping nxapi ssl options when capabilities
are None and generating a warning when these options are skipped

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_nxapi.py:
```diff
@@ -161,9 +197,20 @@ def check_args(module, warnings):
+    os_version = None
+
+    device_info = capabilities.get('device_info')
+    if device_info is not None:
```
@tstoner You can just do `if device_info`? same for os_version and os_platform.

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_nxapi.py:
```diff
@@ -191,8 +238,30 @@ def needs_update(x):
 
-    for parameter in commands.keys():
-        send_commands.append(commands[parameter])
+    if os_platform is not None and os_version is not None:
```
`if os_platform and os_version:`

tstoner committed:
Cleaned up explicit checks for None/not None

trishnaguha commented:
Thanks for the PR @tstoner
